src/storage/v2/disk/storage.cpp

```diff
@@ -603,23 +675,56 @@ VerticesIterable DiskStorage::DiskAccessor::Vertices(LabelId label, PropertyId p
         &transaction_));
   }
 
-  transaction_.index_storage_.emplace_back(std::make_unique<utils::SkipList<storage::Vertex>>());
-  auto &indexed_vertices = transaction_.index_storage_.back();
-  transaction_.index_deltas_storage_.emplace_back();
-  auto &index_deltas = transaction_.index_deltas_storage_.back();
+  utils::SkipList<memgraph::storage::Vertex> *indexed_vertices{nullptr};
+  std::list<storage::Delta> *index_deltas{nullptr};
 
-  auto label_property_filter = [this, &value](const Vertex &vertex, LabelId label, PropertyId property,
-                                              View view) -> bool {
-    return VertexHasLabel(vertex, label, &transaction_, view) &&
-           VertexHasEqualPropertyValue(vertex, property, value, &transaction_, view);
+  auto merge_with_main_cache = [&](auto &index, auto &index_delta_storage) -> std::unordered_set<Gid> {
+    index[cache_key] = utils::SkipList<Vertex>();
+    indexed_vertices = &index[cache_key];
+    index_delta_storage.emplace_back();
+    index_deltas = &index_delta_storage.back();
+
+    auto label_property_filter = [this, &value](const Vertex &vertex, LabelId label, PropertyId property,
+                                                View view) -> bool {
+      return VertexHasLabel(vertex, label, &transaction_, view) &&
+             VertexHasEqualPropertyValue(vertex, property, value, &transaction_, view);
+    };
+
+    return MergeVerticesFromMainCacheWithLabelPropertyIndexCache(label, property, view, *index_deltas,
+                                                                 *indexed_vertices, label_property_filter);
   };
 
-  const auto gids = MergeVerticesFromMainCacheWithLabelPropertyIndexCache(
-      label, property, view, index_deltas, indexed_vertices.get(), label_property_filter);
+  auto &cache = transaction_.label_property_val_index_cache_;
+  auto &cache_ci = transaction_.label_property_val_index_cache_ci_;
+  auto &index_delta_storage = transaction_.index_deltas_storage_;
+
+  SyncDeletedVertices(cache, cache_key, index_delta_storage, view, merge_with_main_cache);
+
+  if (transaction_.command_id > cache_ci) {
+    cache_ci = transaction_.command_id;
+  } else {
+    if (cache.contains(cache_key)) {
+      if (view == View::OLD) {
+        // TODO
+        // we do not need this merge if we can make sure that
+        // the removing and re-adding the same label to a given
+        // vertex within one transaction is not permitted.
+        merge_with_main_cache(cache, index_delta_storage);
+        return VerticesIterable(AllVerticesIterable(cache[cache_key].access(), storage_, &transaction_, view));
+      }
+      cache[cache_key] = utils::SkipList<Vertex>();
+      merge_with_main_cache(cache, index_delta_storage);
```

Comment:
I think it is kinda not expected to go to disk just if you query by different value. IMO this query (with the assumption that label-property index :User(id) exists:
```cypher
MATCH (n:User {id: 10}), (m:User {id: 11})
RETURN n, m;
```
This query should go just once to disk I think. But if you cache based on property value it will go twice. Essentially, this means that we don't get any benefit when doing import